### ISSUE TITLE
Use bike share station in Valhalla 

### ISSUE THREAD

**Issue by xlqian** [valhalla/valhalla#1762]
Hi there, 

Greetings from navitia (again)! 

We're planning to integrate bike sharing into Valhalla so that Valhalla can replace our old bss calculator. 

As we discussed 2 years ago in this issue: https://github.com/valhalla/valhalla/issues/421 (originally created by @TeXitoi), this kind of implementation seemed to be interesting for you. Is that still the case? 

As remind, our requirement is to be able to do "from the start point, walk to the bike share station -> take the bike -> ride the bike until the another bike share station -> put back the bike to the station -> walk to the end point".

After looking around in the code, we found `MultiModalPathAlgorithm` and some commented codes: https://github.com/valhalla/valhalla/blob/master/src/thor/multimodal.cc#L312, from which we guess that you have started the dev but stopped for some reasons, can we have more information about that?

If you're OK with the fact Navitia implements the bike share station in Valhalla, we'll need your support certainly. 

What do you think? 

Thank you & Regards

**Comment by kevinkreiser** [comment]
@xlqian you're right, multimodal is meant to achieve just this usecase with the added caveat that it uses transit time schedule look ups to determine which path to take between "stations". this implementation works only for the multimodal case where you do: walk -> train/bus -> walk -> train/bus -> walk. one could, and we had hoped to, extend this to other modalities such as walk -> bike as you pointed out.

we would be happy if you'd like to contribute more multimodal possibilities to the code base and we'd be happy to give you a hand as you have questions!

**Comment by xlqian** [comment]
Great! Thank you!

**Comment by xlqian** [comment]
After reading the `multimodal.cc`, we realized that the bike sharing was quite different from the transit. 

In the transit, edges of transit live in fact their own graph which intersects with normal graphs(car, bike, pedestrian, etc...) via transit stations. With that said, when it comes to determine if the destination is reached, there won't be problem because we **have to**  switch back to the normal graph via transit stations, there won't be ambiguity of `the node have been already visited, here is the cost, but by which modes?`

For our need (pedestrian->bike->pedestrian, and when a user ask for a this kind of journey, he **doesn't have to take the bike**, for example, his destination is just next to a bike sharing station), the current implementation won't allow us to do so, because of the ambiguity of cost.

(Please correct me if i was wrong. )

Here is some propositions about how to use bike sharing station. 

1. We implement a whole new algorithm only designed for bss:
![image](https://user-images.githubusercontent.com/6093486/56596513-fba59500-65f0-11e9-9c7c-d1fd69b36f3d.png)
  

- We first find the nearest BSS for origin and destination with a distance circle (this distance should be configurable clearly)

- We do a OneToMany on foot from origin to bss around it(respectively, a ManyToOne for destination)

- The we initialize the `edgelabels` with the pedestrian costs computed above and then we lauch a astar/doubleastar with bike...

2. To be continued...

**Comment by kevinkreiser** [comment]
That is a very good distinction in the difference. There is really no need to transition multiple times between modes of travel because unlike transit, you do not have to worry about scheduling and transfering between different bikes. one bike is as good as the next, so the case will be really limited to just `ped->bike->ped`. thanks for the clarification, i must admit i hadn't thought too much about it!

i like the first approaches simplicity. i think we have two options for getting bike sharing stations in the graph. 

1. do some work in the data to attribute certain nodes with a bike_share_station bit. perhaps its enough to just add some fake/false nodes to the graph for this purpose or we could do somethign more similar to what we did with transit and connect edges to a wholly separate node. the big question there is, is this data in OSM
2. if the data is not in OSM or we want to supply our own data at run time we would need to, take the locations as extra parameters in the request and correlate them to the graph on the fly. this is pretty straight forward, we do this for regular locations and avoids, it will just cost more CPU to use this approach.

@gknisely @dnesbitt61 any thoughts?

**Comment by xlqian** [comment]
plop? :)

**Comment by ArthurLeyv** [comment]
Hi @kevinkreiser , do you have any news ? 
Thanks !

**Comment by kevinkreiser** [comment]
Ah sorry, I think I am misunderstanding what is needed from my end. At the moment I dont have any time to devote to actual implementation but I think the general idea that has been laid out seems ok to me. I had assumed that @xlqian would be leading this development?

**Comment by xlqian** [comment]
@kevinkreiser No problem! Yep, If you are OK with that, we'll try out these methods, i'll come back to you certainly for asking for your help! :) thank you!

**Comment by xlqian** [comment]
hello again!

After having implemented the algorithm, I just realized that the bike share stations are not imported[ if they don't belong to an OSMWay](https://github.com/valhalla/valhalla/blob/2f3e28bee048e1b39cb924de01a7524ddc81af5d/src/mjolnir/pbfadminparser.cc#L46) 

So I've two questions about importing a bike sharing station node. (example: https://www.openstreetmap.org/node/6424025687)

1. Importing Nodes (actually can be any node)

It seems that I have to create a new OSM type, let's say OSMNode, and put the nodes that I want into a temporary file like what's done [here](https://github.com/valhalla/valhalla/blob/2345ba1f4936b9b547f1e0e43ba18c1d733a5e52/src/mjolnir/pbfgraphparser.cc#L1544), then set the appropriate properties 

2. Once the node is imported, I should make it appear on the tiles. By that, I mean, project the the node on the graph and create edges if necessary.

Am I right? Do you have any short-cuts? 

thank you

**Comment by kevinkreiser** [comment]
The way we do it with transit is we have a post process which adds edges to existing nodes in the network whose shape follow along the existing edge to the point on the edge where the station would be closest and then a b-line from ther to the station. This seems too heavy duty for your use.

Can we, for your use case, just do a multi point route and use different modes of travel for different legs of the route? To do this we would need to pass in your station locations at run time. This would require we actually know the relevant ones which could technically be not possible but we can make a practical guess of which stations are relevant. in fact, using this approach we could do pretty much all of this logic client side:

1. get the relevant bike stations based on the input coordinates making some assumptions about which ones are too far away to make sense
2. do a pedestrian matrix from the origin location to all the stations
3. do a bike matrix from all stations to each other
4. do a pedestrian matrix from all stations to the destination
5. find the least time permutation of the 3 legs
6. fetch the 3 routes, 1 for each leg with costing as pedestrian, bicycle, pedestrian respectively

Perhaps you already tried this and it didnt fit your performance profile? Or perhaps I am misunderstanding your use-case and this wouldnt handle it? Or maybe there are so so so many stations that we cant pass them at run time?

**Comment by dnesbitt61** [comment]
Kevin's approach is the easiest/quickest way to implement this. The one part that might not be quite correct is that you may want to select the optimal path based on "cost" vs. elapsed time (so bicycle paths favor bike lanes and paths or avoid hills for instance). This means you may want to add cost to the return from a time-distance matrix (or convince @kevinkreiser to add it to Valhalla - seems like a good change it may have some versioning impacts).

I worry that modifications to the path algorithms might be complicated though it would be interesting to experiment and try! You may need to account for paths traversed by bike separately from paths traversed by walking. I am not totally sure, it might work by just tracking the current mode in the EdgeLabels - so long as there is never a case where you have to traverse the same edge by different modes. This could have some failure cases I am not thinking of which would complicate the algorithm. If you choose to try working through this implementation I would start by looking at src/mjolnir/transitbuilder.cc to see how transit stops are connected to edges. A similar method could be applied.

**Comment by xlqian** [comment]
Thank you for your responses. I've been working other things during the last month and just resumed BSS a couple of days ago. I've not yet tested and profiled the algo.
 
@kevinkreiser 's approach is what we have imagined at the first place. easy and quick to implement as you said. 

This biggest disadvantage of this approach, in Navitia's context,  is that it only answers our need of door-to-door journey. We also need to be able to compute a SourceToTarget/TargetToSource matrix in BSS mode from a given coord to, at most, 20000 coords, it's essential for navitia's algo. As you can tell, the performance of matrix will be critical.

@dnesbitt61 `You may need to account for paths traversed by bike separately from paths traversed by walking.` Yes, I was thinking the same thing, that's why I called my second proposed algo **Double Mode Expansion** which still need to be tested and benched

Anyway, we should work on importing BSS first.

**Comment by xlqian** [comment]
Hi again, 

We've been working on a POC project https://github.com/xlqian/TTLS to test our BSS(multimodal) algorithm. Please have a look at the project. I'm working on an another algorithm (Double Expansion).  
I'll come back to you for discussing in details once all algorithms are finished. If they are proved to be OK (both by your team and my team), we'll start implementing in Valhalla. Thank you.

**Comment by kevinkreiser** [comment]
@xlqian i recently added some tracking to bidirectional A* that allows you to make an API call and get back geojson, is there anyway i could modify that to support your visualization directly? see more info here: https://github.com/valhalla/valhalla/pull/1849

**Comment by xlqian** [comment]
Hi, 

It's exactly i'm thinking about. I'd like to be able to use this project as a visual way for better understanding and debugging later. (i was thinking of building python bindings... may be that's a little bit overkill... )
 
In order to visualize the algorithm, the lib that I use (https://github.com/gboeing/osmnx) requires just nodes' OSM Ids.

In case of valhalla, if i'm right, we have no longer nodes' OSM id stored in the tiles, we can export the nodes coordinates in some CSV every N iterations than call a python script to do the rendering, I mean each csv corresponds to one frame.  besides, i'm new to OSMNX too, i don't know whether the lib supports geojson or not, maybe there is a much smatter way to accomplish what we want :)

**Comment by kevinkreiser** [comment]
@xlqian that is correct, we do not have the osmnodeids. What i serialize in my geojson is a new feature for every frame of the animation, where the features geometry is the lat,lon of the osm nodes. right now i only export the nodes at the beginning and ending of the edges (not all nodes in between) to save on space but the code has an option to show full shape as well. my idea was that you could loop over the feature collection in the geojson to visualize it but i never got time to work on it :(

**Comment by xlqian** [comment]
Great, it seams your feature can allow us tovisualize the algo PARTIALLY, how about implementing a parameter like `resolution` or `precision` to have more points?  (how about using google's Encoded Polyline Algorithm Format or something similar to reduce the json size? )

**Comment by xlqian** [comment]
:warning:  the discussion has been diverted :D

**Comment by xlqian** [comment]
A little teasing on the "double expansion a* " algorithm

https://github.com/xlqian/TTLS/blob/master/example_images/double_expansion.gif

The violet dots: Origin and destination 
The blue dots:  BSS inside Paris
The red dots: Exploration on foot 
The Green dots: Exploration on bike

the basic idea is to explore the graph with 4 expansion(2 expansions forward+backward per mode) at the same time.

the advantage of this algo is that:
1. we don't calculate the 1 to N(or N to 1) systematically
2. this algo can tell directly if there is a solution on foot better than a solution on bike... 

I'll switch back to my pull request of importing BSS once I fix a little bug on this algo. sorry for the delay :pray:

**Comment by kevinkreiser** [comment]
@xlqian No worries, take your time, I just wanted to make sure weren't waiting on us for something!

**Comment by xlqian** [comment]
Hello guys, i'm here again. 

After a discussion with our team, I just finished a tiny doc explaining how should we do multimodal pathfinding in Valhalla. Please take a look at this: https://github.com/xlqian/TTLS/blob/master/doc/multimodal_in_valhalla.md

The final solution we adopt is actually quite simple. The results, tested with out minimalist data, seem to be ok.  Being optimistic, i think there won't be a lot of work to do if we implement this method in Valhalla.

**Comment by kevinkreiser** [comment]
@xlqian i just merged the data portion of this work, i cant wait to see the algorithm in action, though the GIFs you made are already really cool! thanks again!

**Comment by chrstnbwnkl** [comment]
This landed quite some time ago
